### ISSUE TITLE
Enhancement/Issue for ActivationSoftmax

### ISSUE THREAD

**Issue by StefanoSamele** [deeplearning4j/deeplearning4j#4008]
This is both a (small) issue and an enhancement for ActivationSoftmax function.
Currently this activation function takes as input only a two dimensional INDArray (we are assuming the dimensions represent batchsize and number of labels), to perform a Softmax operation on each row.
Some architecture, such as SSD, perform both a classification and a location detection: i.e. they output a list of possible boxes with a score for each label. This means that the output has at least three dimensions: (batchsize, number of boxes, number of labels). Still SSD should perform a softmax activation for each box of each batchsize.
We also noticed that you can actually feed to the ActivationSoftmax's method getActivation a 3D input; it thorws no exception and leaves the input untouched (i.e. execAndReturn simply does nothing).
We thought about the following workaround (we already shared it with @maxpumperla ), but we were wondering about the impact on backprop: https://gist.github.com/StefanoSamele/10f08556be7e3f7cffd0ef36a4de714d.

**Comment by shyrma** [comment]
Seems like it is not issue any more, since we implemented softmax as beeing able to take an input array of arbitrary rank ( >= 1), also one has to specify dimension (axis) argument the softmax would be computed along (default is last dimension). One can find corresponding tests here:

https://github.com/deeplearning4j/libnd4j/blob/5aadacb8a72d420ecf70246cddb689c2500b0073/tests_cpu/layers_tests/DeclarableOpsTests1.cpp#L3493-L3620

**Comment by raver119** [comment]
So, in other words new SoftMax op just needs CUDA implementation. Ok.

**Comment by carlthome** [comment]
Status? I'd like to run multi-class semantic segmentation in DL4J but that requires a softmax per pixel.
